tests/JustSaying.TestingFramework/Patiently.cs

```diff
@@ -25,45 +40,64 @@ public static class Patiently
         public static async Task AssertThatAsync(ITestOutputHelper output, Func<Task<bool>> func) =>
             await AssertThatAsync(output, func, 5.Seconds()).ConfigureAwait(false);
 
-        public static async Task AssertThatAsync(ITestOutputHelper output, Func<Task<bool>> func, TimeSpan timeout)
+        public static async Task AssertThatAsync(
+            ITestOutputHelper output,
+            Func<Task<bool>> func,
+            TimeSpan timeout)
         {
-            var started = DateTime.Now;
-            var timeoutAt = DateTime.Now + timeout;
+            var watch = new Stopwatch();
+            watch.Start();
             do
             {
-                if (await func.Invoke().ConfigureAwait(false))
+                try
                 {
-                    return;
+                    if (await func.Invoke().ConfigureAwait(false))
+                    {
+                        return;
+                    }
                 }
+                catch (ShouldAssertException)
```

Comment:
This func is retried in a loop, so each time the assertion fails or the func returns false, we go around again.

When the timeout is complete, we attempt it one last time, and allow that to bubble up so the user sees the correct error. See L71 of the same file.

The effect is that you can put basically anything in that func that returns true or throws an assertion, and it will be retried till it succeeds or times out.